lib/matplotlib/legend.py

```diff
@@ -686,6 +686,19 @@ def _set_artist_props(self, a):
 
         a.set_transform(self.get_transform())
 
+    def set_loc(self, loc):
+        """
+        Set the location of the legend.
+
+        Parameters
+        ----------
+        alignment : {'upper left', 'upper right', 'lower left', 'lower right',
+                    'upper center', 'lower center', 'center left', 'center right'
+                    'center', 'right', 'best'}.
+        """
+        _loc_code = self.codes.get(loc, 'best')
```

Comment:
I'm skeptical of defaulting to `"best"` if an invalid code is given, and in fact think that _all_ of the validation logic from `__init__` (which is actually a pretty significant chunk of code, about 20% of `__init__`) should be moved to this public setter, and `__init__` should just call the public setter (in place of the private setter):

https://github.com/matplotlib/matplotlib/blob/93d288e5b8d0feb69290f2afcab290db77759642/lib/matplotlib/legend.py#L506-L556

And then editing this to use the public constructor:
https://github.com/matplotlib/matplotlib/blob/93d288e5b8d0feb69290f2afcab290db77759642/lib/matplotlib/legend.py#L601-L603

The only thing I slightly question is if there is order dependence in:

- do validation on loc arg
- other setup
- actually set loc

Erroring earlier may preclude some object allocations, etc, and not sure if the location setting has to happen _after_ some of those for some reason (I can't really see why it _would_, the actual private setter is just holding on to some variables and saying its stale... just haven't tested it)

Reply:
Thanks for sugsestion!  I also think it's more reasonable to call the public setter in `__init__`, i will try it later today.

Reply:
Defaulting to `'best'` will actually prevent a good number of the valid values of `loc` from being settable via this method.